topology/src/main/java/io/camunda/zeebe/topology/TopologyInitializer.java

```diff
@@ -0,0 +1,271 @@
+      }
+      return initialized;
+    }
+
+    @Override
+    public void onTopologyUpdated(final ClusterTopology clusterTopology) {
+      if (!clusterTopology.isUninitialized()) {
+        LOGGER.debug("Received cluster topology {} via gossip.", clusterTopology);
+        initialized.complete(true);
+        persistedClusterTopology.removeInitializeListener(this);
+      }
+    }
+  }
+
+  /**
+   * Initializes topology by sending sync requests to other members. If any of them return a valid
+   * topology, it will be initialized. If any of them returns an uninitialized topology, the future
+   * returned by initialize completes as failed.
+   */
+  class SyncInitializer implements TopologyInitializer, Listener {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SyncInitializer.class);
+    private static final Duration SYNC_QUERY_RETRY_DELAY = Duration.ofSeconds(5);
+    private final PersistedClusterTopology persistedClusterTopology;
+    private final ActorFuture<Boolean> initialized;
+    private final List<MemberId> knownMembersToSync;
+    private final ConcurrencyControl executor;
+    private final Function<MemberId, ActorFuture<ClusterTopology>> syncRequester;
+
+    public SyncInitializer(
+        final PersistedClusterTopology persistedClusterTopology,
+        final List<MemberId> knownMembersToSync,
+        final ConcurrencyControl executor,
+        final Function<MemberId, ActorFuture<ClusterTopology>> syncRequester) {
+      this.persistedClusterTopology = persistedClusterTopology;
+      this.knownMembersToSync = knownMembersToSync;
+      this.executor = executor;
+      this.syncRequester = syncRequester;
+      initialized = new CompletableActorFuture<>();
+    }
+
+    @Override
+    public ActorFuture<Boolean> initialize() {
+      if (knownMembersToSync.isEmpty()) {
+        initialized.complete(false);
+      } else {
+        LOGGER.debug("Querying members {} before initializing ClusterTopology", knownMembersToSync);
+        persistedClusterTopology.addUpdateListener(this);
+        onTopologyUpdated(persistedClusterTopology.getTopology());
+        knownMembersToSync.forEach(this::tryInitializeFrom);
+      }
+      return initialized;
+    }
+
+    private void tryInitializeFrom(final MemberId memberId) {
+      requestSync(memberId)
+          .onComplete(
+              (topology, error) -> {
+                if (initialized.isDone()) {
+                  return;
+                }
+                if (error == null && topology != null) {
+                  if (topology.isUninitialized()) {
+                    LOGGER.trace("Cluster topology is uninitialized in {}", memberId);
+                    initialized.complete(false);
+                    return;
+                  }
+                  try {
+                    LOGGER.debug(
+                        "Received cluster topology {} from {}", knownMembersToSync, memberId);
+                    persistedClusterTopology.update(topology);
```

Comment:
:question: I almost missed this: We got a sync response from another member and now we update the persisted topology with it. Shouldn't this be already taken care of by `ClusterTopologyManager#updateLocalTopology` ?
If not, I'd like to see this mentioned in the documentation of `SyncInitializer`.
I think ideally we wouldn't update the persisted topology here, and leave it to `ClusterTopologyManager#initialize`

Reply:
SyncInitializer uses a sync request that bypasses the periodic sync. So it is not updated by ClusterTopologyManager. I find it is simpler to handle this way.
 
For both FileInitializer and GossipInitializer, when completed successfully will have a valid topology in `PersistedClusterTopology`.  To make the semantics consistent, SyncInitializer and StaticInitializer also updates PersistedClusterTopology once it is initialized.